core/model/__init__.py

```diff
@@ -54,6 +54,13 @@ def pg_advisory_lock(
         connection.execute(text(f"SELECT pg_advisory_lock({lock_id});"))
         try:
             yield
+        except IntegrityError:
+            # If there was an IntegrityError, and we are in a transaction,
+            # we need to roll it back before we are able to release the lock.
+            transaction = connection.get_transaction()
+            if transaction is not None:
+                transaction.rollback()
+            raise
```

Comment:
This came up while testing this PR, but it uncovered a bug that existed in our `pg_advisory_lock` function. If something within the lock context manager causes an `IntegrityError` the connection won't accept any other statements until the transaction is rolled back. This means that the `SELECT pg_advisory_unlock` in the `finally` statement will raise an exception, and the lock won't get released. 

In this case we catch the error, and rollback the transaction here, before continuing to raise the error. That way we are able to successfully release the lock.